File: src/solace_autoscale/dispatch/payload.py
"""Pure payload/topic matching primitives for the dispatch rule engine.

No I/O, no clock, no logging — just deterministic functions from (topic, payload, predicate) to a
bool. Payloads are matched by decoding JSON and reading dotted field paths; non-JSON payloads still
match on raw size / prefix so binary workloads are not excluded.

Kept dependency-free on purpose (no ``jsonpath_ng``): a tiny dotted-path reader covers object keys
and list indices, which is all a routing rule needs.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any

#: Operators usable in a payload predicate. ``raw_*`` operate on the undecoded bytes so a rule can
#: match non-JSON payloads (e.g. large binary blobs) by size or byte prefix.
OPERATORS = frozenset({
    "eq", "ne", "in", "nin", "gt", "gte", "lt", "lte",
    "exists", "missing", "prefix", "contains", "regex",
    "raw_size_gt", "raw_size_lt", "raw_prefix",
})

#: Sentinel for "path not present". Shared across the package so ``is`` checks work everywhere.
MISSING = object()
_MISSING = MISSING


class RuleError(ValueError):
    """A rule / predicate that cannot be evaluated as written (bad op, bad regex, ...)."""


def get_path(obj: Any, path: str) -> Any:
    """Read a dotted path out of a decoded JSON value.

    ``order.region`` reads obj["order"]["region"]; a numeric segment indexes a list
    (``items.0.sku``). Returns the ``_MISSING`` sentinel if any segment is absent, so callers can
    distinguish "present and null" from "absent".
    """
    cur = obj
    if path == "":
        return cur
    for seg in path.split("."):
        if isinstance(cur, dict):
            if seg not in cur:
                return _MISSING
            cur = cur[seg]
        elif isinstance(cur, list):
            try:
                idx = int(seg)
            except ValueError:
                return _MISSING
            if idx < 0 or idx >= len(cur):
                return _MISSING
            cur = cur[idx]
        else:
            return _MISSING
    return cur


def decode_json(payload: bytes | str | None) -> Any:
    """Decode a payload to a JSON value, or return ``_MISSING`` if it is not JSON.

    Never raises: a non-JSON payload is a legitimate case handled by ``raw_*`` operators.
    """
    if payload is None:
        return _MISSING
    if isinstance(payload, bytes):
        try:
            payload = payload.decode("utf-8")
        except UnicodeDecodeError:
            return _MISSING
    try:
        return json.loads(payload)
    except (json.JSONDecodeError, ValueError):
        return _MISSING


@dataclass(frozen=True)
class Predicate:
    """One payload condition: ``<path> <op> <value>``.

    For ``raw_*`` operators, ``path`` is ignored and the operator acts on the raw payload bytes.
    ``exists``/``missing`` ignore ``value``.
    """

    path: str
    op: str
    value: Any = None

    def __post_init__(self) -> None:
        if self.op not in OPERATORS:
            raise RuleError(f"unknown operator {self.op!r}; valid: {sorted(OPERATORS)}")
        if self.op == "regex":
            try:
                re.compile(str(self.value))
            except re.error as e:
                raise RuleError(f"invalid regex {self.value!r}: {e}") from e


def _as_bytes(payload: bytes | str | None) -> bytes:
    if payload is None:
        return b""
    return payload if isinstance(payload, bytes) else payload.encode("utf-8")


def eval_predicate(pred: Predicate, decoded: Any, payload: bytes | str | None) -> bool:
    """Evaluate a single predicate. Pure. Missing fields are False for every op except ``missing``."""
    op = pred.op

    if op in ("raw_size_gt", "raw_size_lt", "raw_prefix"):
        raw = _as_bytes(payload)
        if op == "raw_size_gt":
            return len(raw) > int(pred.value)
        if op == "raw_size_lt":
            return len(raw) < int(pred.value)
        # raw_prefix
        pref = pred.value
        pref_b = pref.encode("utf-8") if isinstance(pref, str) else bytes(pref)
        return raw.startswith(pref_b)

    actual = get_path(decoded, pred.path)

    if op == "exists":
        return actual is not _MISSING
    if op == "missing":
        return actual is _MISSING
    if actual is _MISSING:
        return False  # any comparison against an absent field is False

    if op == "eq":
        return bool(actual == pred.value)
    if op == "ne":
        return bool(actual != pred.value)
    if op == "in":
        return actual in _iterable(pred.value)
    if op == "nin":
        return actual not in _iterable(pred.value)
    if op in ("gt", "gte", "lt", "lte"):
        return _compare(op, actual, pred.value)
    if op == "prefix":
        return isinstance(actual, str) and actual.startswith(str(pred.value))
    if op == "contains":
        if isinstance(actual, str):
            return str(pred.value) in actual
        if isinstance(actual, list):
            return pred.value in actual
        return False
    if op == "regex":
        return isinstance(actual, str) and re.search(str(pred.value), actual) is not None
    raise RuleError(f"unhandled operator {op!r}")  # pragma: no cover


def _iterable(value: Any) -> Any:
    return value if isinstance(value, (list, tuple, set)) else (value,)


def _compare(op: str, a: Any, b: Any) -> bool:
    try:
        if op == "gt":
            return bool(a > b)
        if op == "gte":
            return bool(a >= b)
        if op == "lt":
            return bool(a < b)
        return bool(a <= b)  # lte
    except TypeError:
        return False  # e.g. comparing str to int — a rule that does not apply, not an error


def topic_matches(pattern: str, topic: str) -> bool:
    """Match a topic against a Solace-style pattern using ``*`` (one level) and ``>`` (rest).

    Levels are ``/``-separated. ``*`` matches exactly one level; ``>`` matches this level and every
    level after it and must be the final token. A pattern with no wildcards is an exact match.
    """
    if pattern == topic:
        return True
    pat = pattern.split("/")
    top = topic.split("/")
    for i, seg in enumerate(pat):
        if seg == ">":
            return i < len(top)  # matches one-or-more remaining levels
        if i >= len(top):
            return False
        if seg == "*":
            continue
        if seg != top[i]:
            return False
    return len(pat) == len(top)

File: shim/rules/payload.go
package rules

import (
	"encoding/json"
	"reflect"
	"regexp"
	"strconv"
	"strings"
)

// Operators usable in a payload predicate. raw_* operate on the undecoded bytes so a rule can match
// non-JSON payloads (e.g. large binary blobs) by size or byte prefix.
var operators = map[string]bool{
	"eq": true, "ne": true, "in": true, "nin": true,
	"gt": true, "gte": true, "lt": true, "lte": true,
	"exists": true, "missing": true, "prefix": true, "contains": true, "regex": true,
	"raw_size_gt": true, "raw_size_lt": true, "raw_prefix": true,
}

// missing is the sentinel for "path not present". A distinct pointer value so it can be compared by
// identity and never collides with a real decoded JSON value (which is never this type).
type missingT struct{}

var missing = missingT{}

func isMissing(v any) bool {
	_, ok := v.(missingT)
	return ok
}

// getPath reads a dotted path out of a decoded JSON value. "order.region" reads
// obj["order"]["region"]; a numeric segment indexes a list ("items.0.sku"). Returns the missing
// sentinel if any segment is absent, so callers can distinguish "present and null" from "absent".
func getPath(obj any, path string) any {
	cur := obj
	if path == "" {
		return cur
	}
	for _, seg := range strings.Split(path, ".") {
		switch c := cur.(type) {
		case map[string]any:
			next, ok := c[seg]
			if !ok {
				return missing
			}
			cur = next
		case []any:
			idx, err := strconv.Atoi(seg)
			if err != nil || idx < 0 || idx >= len(c) {
				return missing
			}
			cur = c[idx]
		default:
			return missing
		}
	}
	return cur
}

// decodeJSON decodes a payload to a JSON value, or returns the missing sentinel if it is not JSON.
// Never errors: a non-JSON payload is a legitimate case handled by the raw_* operators. Numbers
// decode to json.Number so integer comparisons stay exact.
func decodeJSON(payload []byte) any {
	if payload == nil {
		return missing
	}
	dec := json.NewDecoder(strings.NewReader(string(payload)))
	dec.UseNumber()
	var v any
	if err := dec.Decode(&v); err != nil {
		return missing
	}
	// Reject trailing content (matches Python json.loads, which parses a single value).
	if dec.More() {
		return missing
	}
	return v
}

// Predicate is one payload condition: <Path> <Op> <Value>. For raw_* operators Path is ignored and
// the operator acts on the raw payload bytes. exists/missing ignore Value.
type Predicate struct {
	Path  string
	Op    string
	Value any
}

// evalPredicate evaluates a single predicate. Pure. Missing fields are false for every op except
// "missing". A type-mismatched comparison (e.g. number vs string) is false, not an error.
func evalPredicate(p Predicate, decoded any, payload []byte) bool {
	switch p.Op {
	case "raw_size_gt":
		return len(payload) > toInt(p.Value)
	case "raw_size_lt":
		return len(payload) < toInt(p.Value)
	case "raw_prefix":
		return strings.HasPrefix(string(payload), toStr(p.Value))
	}

	actual := getPath(decoded, p.Path)

	switch p.Op {
	case "exists":
		return !isMissing(actual)
	case "missing":
		return isMissing(actual)
	}
	if isMissing(actual) {
		return false // any comparison against an absent field is false
	}

	switch p.Op {
	case "eq":
		return jsonEqual(actual, p.Value)
	case "ne":
		return !jsonEqual(actual, p.Value)
	case "in":
		return containsValue(p.Value, actual)
	case "nin":
		return !containsValue(p.Value, actual)
	case "gt", "gte", "lt", "lte":
		return compare(p.Op, actual, p.Value)
	case "prefix":
		s, ok := actual.(string)
		return ok && strings.HasPrefix(s, toStr(p.Value))
	case "contains":
		switch a := actual.(type) {
		case string:
			return strings.Contains(a, toStr(p.Value))
		case []any:
			return containsValue(a, p.Value)
		}
		return false
	case "regex":
		s, ok := actual.(string)
		if !ok {
			return false
		}
		re, err := regexp.Compile(toStr(p.Value))
		if err != nil {
			return false
		}
		return re.FindStringIndex(s) != nil
	}
	return false
}

// jsonEqual compares two decoded/spec JSON values for equality, treating numbers by value across the
// json.Number / float64 / int spelling differences between a decoded payload and a spec literal.
func jsonEqual(a, b any) bool {
	if an, aok := asFloat(a); aok {
		if bn, bok := asFloat(b); bok {
			return an == bn
		}
		return false
	}
	return reflect.DeepEqual(a, b)
}

// containsValue reports whether haystack (a list, or a scalar treated as a one-element list) contains
// needle, using jsonEqual for element comparison.
func containsValue(haystack, needle any) bool {
	switch h := haystack.(type) {
	case []any:
		for _, el := range h {
			if jsonEqual(el, needle) {
				return true
			}
		}
		return false
	default:
		return jsonEqual(haystack, needle)
	}
}

// compare evaluates gt/gte/lt/lte for numbers and strings. A cross-type comparison is false.
func compare(op string, a, b any) bool {
	if af, aok := asFloat(a); aok {
		if bf, bok := asFloat(b); bok {
			return cmpOrder(op, af > bf, af >= bf, af < bf, af <= bf)
		}
		return false
	}
	if as, aok := a.(string); aok {
		if bs, bok := b.(string); bok {
			return cmpOrder(op, as > bs, as >= bs, as < bs, as <= bs)
		}
		return false
	}
	return false
}

func cmpOrder(op string, gt, gte, lt, lte bool) bool {
	switch op {
	case "gt":
		return gt
	case "gte":
		return gte
	case "lt":
		return lt
	default: // lte
		return lte
	}
}

// asFloat coerces a decoded JSON number (json.Number), a spec literal (float64/int), or a numeric
// bool-free scalar to float64. Returns ok=false for non-numeric values (including bool).
func asFloat(v any) (float64, bool) {
	switch n := v.(type) {
	case json.Number:
		f, err := n.Float64()
		return f, err == nil
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	}
	return 0, false
}

func toInt(v any) int {
	if f, ok := asFloat(v); ok {
		return int(f)
	}
	if s, ok := v.(string); ok {
		if i, err := strconv.Atoi(s); err == nil {
			return i
		}
	}
	return 0
}

func toStr(v any) string {
	switch s := v.(type) {
	case string:
		return s
	case json.Number:
		return s.String()
	case nil:
		return ""
	default:
		return ""
	}
}
